### PR TITLE
GODRIVER-2741 Do not create or drop `eccCollection`

### PR DESCRIPTION
**Pull request by kevinAlbs**
GODRIVER-2741
# Summary

- Do not create or drop `eccCollection`
- Return error if maxWireVersion < 21 when creating a Queryable Encryption (QE) collection.
- Support `null` in comparison of command monitoring events

## Other improvements
- Include Expected and Got in command monitoring match errors

# Background & Motivation

See DRIVERS-2524 for additional background behind these changes.

maxWireVersion=21 is for server 7.0.

### Include Expected and Got in command monitoring match errors

Before this PR, here is an example error from an event mismatch:

```
=== RUN   TestClientSideEncryptionSpec
=== RUN   TestClientSideEncryptionSpec/fle2v2-CreateCollection.json
=== RUN   TestClientSideEncryptionSpec/fle2v2-CreateCollection.json/state_collections_and_index_are_created
    unified_spec_test.go:346: 
                Error Trace:    /home/ubuntu/code/tasks/mongo-go-driver-G2741/mongo/integration/cmd_monitoring_helpers_test.go:269
                                                        /home/ubuntu/code/tasks/mongo-go-driver-G2741/mongo/integration/unified_spec_test.go:346
                                                        /home/ubuntu/code/tasks/mongo-go-driver-G2741/mongo/integration/mongotest.go:263
                Error:          Should be true
                Test:           TestClientSideEncryptionSpec/fle2v2-CreateCollection.json/state_collections_and_index_are_created
                Messages:       expectation comparison error at index 0: value mismatch for key drop; expected !enxcol_.encryptedCollection.esc2 (hex=21000000656e78636f6c5f2e656e63727970746564436f6c6c656374696f6e2e6573633200), got  enxcol_.encryptedCollection.esc (hex=20000000656e78636f6c5f2e656e63727970746564436f6c6c656374696f6e2e65736300)
--- FAIL: TestClientSideEncryptionSpec (0.07s)
    --- FAIL: TestClientSideEncryptionSpec/fle2v2-CreateCollection.json (0.07s)
        --- FAIL: TestClientSideEncryptionSpec/fle2v2-CreateCollection.json/state_collections_and_index_are_created (0.06s)
FAIL
FAIL    go.mongodb.org/mongo-driver/mongo/integration   0.079s
FAIL
```

The addition of `makeMatchError` is intended to show the `Expected` and `Got` events in the error. Example:

```
=== RUN   TestClientSideEncryptionSpec
=== RUN   TestClientSideEncryptionSpec/fle2v2-CreateCollection.json
=== RUN   TestClientSideEncryptionSpec/fle2v2-CreateCollection.json/state_collections_and_index_are_created
    unified_spec_test.go:346: 
                Error Trace:    /home/ubuntu/code/tasks/mongo-go-driver-G2741/mongo/integration/cmd_monitoring_helpers_test.go:269
                                                        /home/ubuntu/code/tasks/mongo-go-driver-G2741/mongo/integration/unified_spec_test.go:346
                                                        /home/ubuntu/code/tasks/mongo-go-driver-G2741/mongo/integration/mongotest.go:263
                Error:          Expected nil, but got: &errors.errorString{s:"value mismatch for key drop; expected !\x00\x00\x00enxcol_.encryptedCollection.esc2\x00 (hex=21000000656e78636f6c5f2e656e63727970746564436f6c6c656374696f6e2e6573633200), got  \x00\x00\x00enxcol_.encryptedCollection.esc\x00 (hex=20000000656e78636f6c5f2e656e63727970746564436f6c6c656374696f6e2e65736300)\nExpected {\"drop\":\"enxcol_.encryptedCollection.esc2\"}\nGot: {\"drop\":\"enxcol_.encryptedCollection.esc\",\"lsid\":{\"id\":{\"$binary\":{\"base64\":\"IRJwfhl1RDy5D4/hlUGZ+w==\",\"subType\":\"04\"}}},\"$clusterTime\":{\"clusterTime\":{\"$timestamp\":{\"t\":1682351987,\"i\":1}},\"signature\":{\"hash\":{\"$binary\":{\"base64\":\"AAAAAAAAAAAAAAAAAAAAAAAAAAA=\",\"subType\":\"00\"}},\"keyId\":{\"$numberLong\":\"0\"}}},\"$db\":\"default\"}"}
                Test:           TestClientSideEncryptionSpec/fle2v2-CreateCollection.json/state_collections_and_index_are_created
                Messages:       expectation comparison error at index 0: value mismatch for key drop; expected !enxcol_.encryptedCollection.esc2 (hex=21000000656e78636f6c5f2e656e63727970746564436f6c6c656374696f6e2e6573633200), got  enxcol_.encryptedCollection.esc (hex=20000000656e78636f6c5f2e656e63727970746564436f6c6c656374696f6e2e65736300)
                                Expected {"drop":"enxcol_.encryptedCollection.esc2"}
                                Got: {"drop":"enxcol_.encryptedCollection.esc","lsid":{"id":{"$binary":{"base64":"IRJwfhl1RDy5D4/hlUGZ+w==","subType":"04"}}},"$clusterTime":{"clusterTime":{"$timestamp":{"t":1682351987,"i":1}},"signature":{"hash":{"$binary":{"base64":"AAAAAAAAAAAAAAAAAAAAAAAAAAA=","subType":"00"}},"keyId":{"$numberLong":"0"}}},"$db":"default"}
--- FAIL: TestClientSideEncryptionSpec (0.08s)
    --- FAIL: TestClientSideEncryptionSpec/fle2v2-CreateCollection.json (0.07s)
        --- FAIL: TestClientSideEncryptionSpec/fle2v2-CreateCollection.json/state_collections_and_index_are_created (0.07s)
FAIL
FAIL    go.mongodb.org/mongo-driver/mongo/integration   0.088s
FAIL
```

## Support `null` in comparison of command monitoring events

Supporting `null` in comparison of command monitoring events is a requirement of the [Transactions test format](https://github.com/mongodb/specifications/tree/master/source/transactions/tests#null-values):

> Tests MUST assert that the actual command omits any field that has a `null` value in the expected command.

The legacy Client-Side Encryption tests are [based on the Transactions test format](https://github.com/mongodb/specifications/blob/master/source/client-side-encryption/tests/README.rst#spec-test-format):

> The spec tests format is an extension of transactions spec tests with some additions

### CODE REVIEW

File: mongo/integration/cmd_monitoring_helpers_test.go [hunk 1]
```diff
@@ -180,6 +181,16 @@ func compareDocsHelper(mt *mtest.T, expected, actual bson.Raw, prefix string) er
...
+		if e.Value().Type == bson.TypeNull {
+			// Expected value is BSON null. Expect the actual field to be omitted.
+			if err == bsoncore.ErrElementNotFound {
+				continue
+			} else if err != nil {
+				return fmt.Errorf("expected key %q to be omitted but got error: %v", eKey, err)
+			} else {
+				return fmt.Errorf("expected key %q to be omitted but got %q", eKey, aVal)
+			}
+		}
```
Comment (prestonvasquez) [inline mongo/integration/cmd_monitoring_helpers_test.go:193]:
```suggestion
			// Expected value is BSON null. Expect the actual field to be omitted.
			if errors.Is(err, bsoncore.ErrElementNotFound) {
				continue
			}

			if err != nil {
				return fmt.Errorf("expected key %q to be omitted but got error: %v", eKey, err)
			}

			return fmt.Errorf("expected key %q to be omitted but got %q", eKey, aVal)
```

This suggestion:

(1) removes the if/else branching
(2) uses the `errors` package to check if the error is `bsoncore.ErrElementNotFound`

File: mongo/integration/cmd_monitoring_helpers_test.go [hunk 2]
```diff
@@ -293,8 +315,14 @@ func compareStartedEvent(mt *mtest.T, expectation *expectation, id0, id1 bson.Ra
...
-			assert.NotNil(mt, err, "expected key %q to be omitted but got %q", key, actualVal)
-			continue
+			// Expected value is BSON null. Expect the actual field to be omitted.
+			if err == bsoncore.ErrElementNotFound {
+				continue
+			} else if err != nil {
+				return makeMatchError(mt, expected.Command, evt.Command, "expected key %q to be omitted but got error: %v", key, err)
+			} else {
+				return makeMatchError(mt, expected.Command, evt.Command, "expected key %q to be omitted but got %q", key, actualVal)
+			}
```
Comment (prestonvasquez) [inline mongo/integration/cmd_monitoring_helpers_test.go:325]:
```suggestion
			// Expected value is BSON null. Expect the actual field to be omitted.
			if errors.Is(err, bsoncore.ErrElementNotFound) {
				continue
			}

			if err != nil {
				return makeMatchError(mt, expected.Command, evt.Command, "expected key %q to be omitted but got error: %v", key, err)
			}

			return makeMatchError(mt, expected.Command, evt.Command, "expected key %q to be omitted but got %q", key, actualVal)

```


			
This suggestion:

(1) removes the if/else branching
(2) uses the `errors` package to check if the error is `bsoncore.ErrElementNotFound`

File: mongo/integration/cmd_monitoring_helpers_test.go [hunk 3]
```diff
@@ -259,6 +270,17 @@ func checkExpectations(mt *mtest.T, expectations *[]*expectation, id0, id1 bson.
 	}
 }
 
+// makeMatchError appends `expected` and `actual` BSON data to an error.
+func makeMatchError(mt *mtest.T, expected bson.Raw, actual bson.Raw, format string, args ...interface{}) error {
```
Comment (prestonvasquez) [inline mongo/integration/cmd_monitoring_helpers_test.go:274]: Could we call this function `newMatchError` instead? `make*` is often used for functions that allocate and initialize , something like this:

```go
func makeIntArray(size int, val int) []int {
    result := make([]int, size)
    for i := range result {
        result[i] = initialValue 
    }

   return result
}
```

here are the make docs: https://go.dev/ref/spec#Making_slices_maps_and_channels

`new` allocates a memory for a new value and returns a pointer to it, which is what `makeMatchError` is doing. See the fmt.Errorf function for the pointer that is returned: https://cs.opensource.google/go/go/+/refs/tags/go1.20.3:src/fmt/errors.go;l=22

File: mongo/database.go [hunk 4]
```diff
@@ -579,7 +579,27 @@ func (db *Database) createCollectionWithEncryptedFields(ctx context.Context, nam
...
 
-	// Create the three encryption-related, associated collections: `escCollection`, `eccCollection` and `ecocCollection`.
+	// Check the wire version to ensure server is 7.0.0 or newer.
+	// After the wire version check, and before creating the collections, it is possible the server state changes.
+	// That is OK. This wire version check is a best effort to inform users earlier if using a QEv2 driver with a QEv1 server.
+	{
+		const QEv2WireVersion = 21
+		server, err := db.client.deployment.SelectServer(ctx, description.WriteSelector())
+		if err != nil {
+			return fmt.Errorf("error selecting server to check maxWireVersion: %s", err)
```
Comment (prestonvasquez) [inline mongo/database.go:589]:
```suggestion
			return fmt.Errorf("error selecting server to check maxWireVersion: %w", err)
```

File: mongo/database.go [hunk 5]
```diff
@@ -579,7 +579,27 @@ func (db *Database) createCollectionWithEncryptedFields(ctx context.Context, nam
...
+		}
+		conn, err := server.Connection(ctx)
+		if err != nil {
+			return fmt.Errorf("error getting connection to check maxWireVersion: %s", err)
```
Comment (prestonvasquez) [inline mongo/database.go:593]:
```suggestion
			return fmt.Errorf("error getting connection to check maxWireVersion: %w", err)
```

File: mongo/integration/cmd_monitoring_helpers_test.go [hunk 6]
```diff
@@ -180,6 +181,16 @@ func compareDocsHelper(mt *mtest.T, expected, actual bson.Raw, prefix string) er
 		}
 
 		aVal, err := actual.LookupErr(eKey)
+		if e.Value().Type == bson.TypeNull {
+			// Expected value is BSON null. Expect the actual field to be omitted.
+			if errors.Is(err, bsoncore.ErrElementNotFound) {
+				continue
+			}
+			if err != nil {
+				return fmt.Errorf("expected key %q to be omitted but got error: %v", eKey, err)
```
Comment (prestonvasquez) [inline mongo/integration/cmd_monitoring_helpers_test.go:190]:
```suggestion
				return fmt.Errorf("expected key %q to be omitted but got error: %w", eKey, err)
```

**Review by prestonvasquez — APPROVED** [review]
LGTM

**Review by qingyang-hu — APPROVED** [review]
👍

**Pull request merged**